cpp/src/arrow/compute/kernels/scalar_compare.cc

```diff
@@ -385,6 +385,23 @@ struct VarArgsCompareFunction : ScalarFunction {
   }
 };
 
+Result<TypeHolder> ResolveDecimalCompareOutputType(KernelContext*,
+                                                   const std::vector<TypeHolder>& types) {
+  // casted types should be same size decimals
+  const auto& left_type = checked_cast<const DecimalType&>(*types[0]);
+  const auto& right_type = checked_cast<const DecimalType&>(*types[1]);
+  DCHECK_EQ(left_type.id(), right_type.id());
+
+  // check the casted decimal scales according kAdd promotion rule
+  const int32_t s1 = left_type.scale();
+  const int32_t s2 = right_type.scale();
+  if (s1 != s2) {
+    return Status::Invalid("Comparison of two decimal  ", "types s1 != s2. (", s1, s2,
+                           ").");
+  }
+  return boolean();
+}
+
```

Comment:
It's an interesting approach to use the output-type resolver to assert on the input types, but what you really doing here is asserting on the inputs. I think you should write a custom matcher instead. Compare always returns `boolean()`, there is nothing to resolve.

That would also help (in the future) dispatching to kernels that can deal with different scales dynamically. Values can still be logically equal even though they are represented physically in memory with different scales.

Reply:
I've read the issue and it seems that the issue here is that we need a special cast when the scales don't match because the generic cast can't currently handle that.

Reply:
Or maybe we want users to be explicit with casts before comparing. I still think this should be handled by the input matching machinery.

Reply:
> but what you really doing here is asserting on the inputs.

Yes, this line of code `DCHECK_EQ(left_type.id(), right_type.id());` in `ResolveDecimalCompareOutputType` is problematic because we might be comparing `float and decimal` or `decimal128 and decimal256`...


> Or maybe we want users to be explicit with casts before comparing. I still think this should be handled by the input matching machinery.

The current matcher system only works on matching of builtin types, and what the compare kernel function needs is whether the dependencies between builtin types meet the requirements. This requires adding a `bool Matches(const std::vector<TypeHolder>& types) const;` interface  to `TypeMatcher`.

That means the previous semantics of `TypeMatcher` was to check the validity of builtin types. Now it is necessary to check the legality of dependencies generated when functions use builtin types. Is this reasonable for the design of TypeMatcher?

But in fact, it is indeed more reasonable to use matcher to determine whether the comparison operation of decimal types is legal.

Reply:
Besides, there is a similar case in `IfElse` related kernel functions : https://github.com/apache/arrow/pull/41363.
Do the input types' check in the output `resolver` in `IfElse` case is reasonable?

Reply:
> Is this reasonable for the design of TypeMatcher?

As long as it checks only the types and the code is simple enough that we can reason about the set of combinations handled.

If some combinations are invalid, but you want to generate a message, I think you could have the type matcher route bad combinations to these fail kernels that produce an error instead of automatically casting inputs.

Reply:
That way, the output type of "compare" remains boolean as it should be. No extra logic needed for output type resolution, but the function is not total — for some input types, it produces a bad `Status`.

Reply:
> As long as it checks only the types and the code is simple enough that we can reason about the set of combinations handled.

Thanks, it's more concise to use `TypeMatcher` check the input types.